File: backend/fastapi_server/app/models/rag.py
"""Data models for RAG corpus management."""

from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum
from pydantic import BaseModel, Field, validator
from uuid import UUID, uuid4


class UploadStatus(str, Enum):
    """Upload status enumeration."""
    QUEUED = "queued"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class ChunkingStrategy(str, Enum):
    """Document chunking strategy enumeration."""
    AUTO = "auto"
    FIXED = "fixed"
    SEMANTIC = "semantic"


class CorpusCreateRequest(BaseModel):
    """Request model for creating a new corpus."""
    
    display_name: str = Field(..., min_length=1, max_length=100, description="Human-readable corpus name")
    description: Optional[str] = Field(None, max_length=500, description="Corpus description")
    chatbot_id: str = Field(..., description="Associated chatbot ID")
    embedding_model: str = Field(default="publishers/google/models/text-embedding-004", description="Embedding model to use")
    
    @validator('display_name')
    def validate_display_name(cls, v):
        if not v.strip():
            raise ValueError('Display name cannot be empty or only whitespace')
        return v.strip()


class CorpusUpdateRequest(BaseModel):
    """Request model for updating corpus metadata."""
    
    display_name: Optional[str] = Field(None, min_length=1, max_length=100, description="Updated corpus name")
    description: Optional[str] = Field(None, max_length=500, description="Updated corpus description")
    
    @validator('display_name')
    def validate_display_name(cls, v):
        if v is not None and not v.strip():
            raise ValueError('Display name cannot be empty or only whitespace')
        return v.strip() if v else v


class CorpusResponse(BaseModel):
    """Response model for corpus information."""
    
    id: UUID = Field(..., description="Corpus UUID")
    vertex_corpus_name: str = Field(..., description="Vertex AI corpus resource name")
    display_name: str = Field(..., description="Human-readable corpus name")
    description: Optional[str] = Field(None, description="Corpus description")
    chatbot_id: str = Field(..., description="Associated chatbot ID")
    document_count: int = Field(default=0, description="Number of documents in corpus")
    total_size_bytes: int = Field(default=0, description="Total size of all documents")
    embedding_model: str = Field(..., description="Embedding model used")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")


class DocumentUploadRequest(BaseModel):
    """Request model for document upload."""
    
    corpus_id: UUID = Field(..., description="Target corpus ID")
    display_name: Optional[str] = Field(None, max_length=200, description="Custom display name for document")
    description: Optional[str] = Field(None, max_length=500, description="Document description")
    chunking_strategy: ChunkingStrategy = Field(default=ChunkingStrategy.AUTO, description="Chunking strategy")
    metadata: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Additional metadata")


class DocumentResponse(BaseModel):
    """Response model for document information."""
    
    id: UUID = Field(..., description="Document UUID")
    corpus_id: UUID = Field(..., description="Parent corpus ID")
    vertex_file_name: str = Field(..., description="Vertex AI file resource name")
    original_filename: str = Field(..., description="Original filename")
    display_name: str = Field(..., description="Display name")
    file_size_bytes: int = Field(..., description="File size in bytes")
    mime_type: str = Field(..., description="MIME type")
    upload_status: UploadStatus = Field(..., description="Upload processing status")
    error_message: Optional[str] = Field(None, description="Error message if upload failed")
    metadata: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Additional metadata")
    created_at: datetime = Field(..., description="Upload timestamp")
    processed_at: Optional[datetime] = Field(None, description="Processing completion timestamp")


class BatchUploadRequest(BaseModel):
    """Request model for batch document upload."""
    
    corpus_id: UUID = Field(..., description="Target corpus ID")
    chunking_strategy: ChunkingStrategy = Field(default=ChunkingStrategy.AUTO, description="Chunking strategy")
    metadata: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Additional metadata for all files")


class UploadJobResponse(BaseModel):
    """Response model for upload job status."""
    
    id: UUID = Field(..., description="Upload job UUID")
    corpus_id: UUID = Field(..., description="Target corpus ID")
    status: UploadStatus = Field(..., description="Overall job status")
    total_files: int = Field(..., description="Total number of files to process")
    processed_files: int = Field(default=0, description="Number of files processed")
    successful_uploads: int = Field(default=0, description="Number of successful uploads")
    failed_uploads: int = Field(default=0, description="Number of failed uploads")
    progress_percentage: float = Field(default=0.0, description="Progress percentage (0-100)")
    estimated_time_remaining: Optional[int] = Field(None, description="Estimated time remaining in seconds")
    errors: List[str] = Field(default_factory=list, description="Error messages")
    created_at: datetime = Field(..., description="Job creation timestamp")
    started_at: Optional[datetime] = Field(None, description="Processing start timestamp")
    completed_at: Optional[datetime] = Field(None, description="Job completion timestamp")
    
    @property
    def is_active(self) -> bool:
        """Check if upload job is still active."""
        return self.status in [UploadStatus.QUEUED, UploadStatus.PROCESSING]
    
    @property
    def is_completed(self) -> bool:
        """Check if upload job is completed (success or failure)."""
        return self.status in [UploadStatus.COMPLETED, UploadStatus.FAILED, UploadStatus.CANCELLED]


class FileValidationError(BaseModel):
    """File validation error details."""
    
    filename: str = Field(..., description="Name of the problematic file")
    error_type: str = Field(..., description="Type of validation error")
    message: str = Field(..., description="Detailed error message")


class UploadValidationResponse(BaseModel):
    """Response model for upload validation."""
    
    valid_files: List[str] = Field(default_factory=list, description="List of valid filenames")
    invalid_files: List[FileValidationError] = Field(default_factory=list, description="List of validation errors")
    total_size_bytes: int = Field(..., description="Total size of valid files")
    estimated_processing_time: Optional[int] = Field(None, description="Estimated processing time in seconds")
    
    @property
    def is_valid(self) -> bool:
        """Check if all files passed validation."""
        return len(self.invalid_files) == 0 and len(self.valid_files) > 0


class CorpusStatsResponse(BaseModel):
    """Response model for corpus statistics."""
    
    corpus_id: UUID = Field(..., description="Corpus UUID")
    document_count: int = Field(..., description="Number of documents")
    total_size_bytes: int = Field(..., description="Total size of all documents")
    total_chunks: Optional[int] = Field(None, description="Total number of text chunks")
    average_chunk_size: Optional[float] = Field(None, description="Average chunk size in characters")
    embedding_dimensions: Optional[int] = Field(None, description="Embedding vector dimensions")
    last_updated: datetime = Field(..., description="Last corpus update timestamp")
    supported_file_types: List[str] = Field(default_factory=list, description="Supported file MIME types")


class DocumentSearchRequest(BaseModel):
    """Request model for searching documents within a corpus."""
    
    corpus_id: UUID = Field(..., description="Corpus to search in")
    query: str = Field(..., min_length=1, max_length=500, description="Search query")
    max_results: int = Field(default=10, ge=1, le=50, description="Maximum number of results")
    min_relevance_score: float = Field(default=0.0, ge=0.0, le=1.0, description="Minimum relevance score")
    file_types: Optional[List[str]] = Field(None, description="Filter by file types")
    
    @validator('query')
    def validate_query(cls, v):
        if not v.strip():
            raise ValueError('Search query cannot be empty or only whitespace')
        return v.strip()


class DocumentSearchResult(BaseModel):
    """Individual document search result."""
    
    document_id: UUID = Field(..., description="Document UUID")
    filename: str = Field(..., description="Document filename")
    relevance_score: float = Field(..., description="Relevance score (0.0 to 1.0)")
    snippet: str = Field(..., description="Relevant content snippet")
    chunk_index: Optional[int] = Field(None, description="Index of the matching chunk")
    metadata: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Document metadata")


class DocumentSearchResponse(BaseModel):
    """Response model for document search."""
    
    corpus_id: UUID = Field(..., description="Searched corpus ID")
    query: str = Field(..., description="Original search query")
    results: List[DocumentSearchResult] = Field(default_factory=list, description="Search results")
    total_results: int = Field(..., description="Total number of matching results")
    processing_time_ms: Optional[int] = Field(None, description="Search processing time")

File: frontend/types/rag.ts
/**
 * TypeScript interfaces for RAG (Retrieval Augmented Generation) management
 * These interfaces match the backend API response models
 */

export type UploadStatus = 'queued' | 'processing' | 'completed' | 'failed' | 'cancelled';
export type ChunkingStrategy = 'auto' | 'fixed' | 'semantic';

// Core RAG entities
export interface Corpus {
  id: string;
  vertex_corpus_name: string;
  display_name: string;
  description?: string;
  chatbot_id: string;
  document_count: number;
  total_size_bytes: number;
  embedding_model: string;
  created_at: string;
  updated_at: string;
}

export interface Document {
  id: string;
  corpus_id: string;
  vertex_file_name: string;
  original_filename: string;
  display_name: string;
  file_size_bytes: number;
  mime_type: string;
  file_type?: string;
  upload_status: UploadStatus;
  error_message?: string;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
  processed_at?: string;
}

export interface UploadJob {
  id: string;
  corpus_id: string;
  status: UploadStatus;
  total_files: number;
  processed_files: number;
  successful_uploads: number;
  failed_uploads: number;
  progress_percentage: number;
  estimated_time_remaining?: number;
  errors: string[];
  created_at: string;
  started_at?: string;
  completed_at?: string;
}

// API Request types
export interface CreateCorpusRequest {
  display_name: string;
  description?: string;
  chatbot_id: string;
  embedding_model?: string;
}

export interface UpdateCorpusRequest {
  display_name?: string;
  description?: string;
}

export interface DocumentUploadRequest {
  corpus_id: string;
  display_name?: string;
  description?: string;
  chunking_strategy?: ChunkingStrategy;
  metadata?: Record<string, any>;
}

export interface BatchUploadRequest {
  corpus_id: string;
  chunking_strategy?: ChunkingStrategy;
  metadata?: Record<string, any>;
}

export interface DocumentSearchRequest {
  corpus_id: string;
  query: string;
  max_results?: number;
  min_relevance_score?: number;
  file_types?: string[];
}

// API Response types
export interface FileValidationError {
  filename: string;
  error_type: string;
  message: string;
}

export interface UploadValidationResponse {
  valid_files: string[];
  invalid_files: FileValidationError[];
  total_size_bytes: number;
  estimated_processing_time?: number;
}

export interface DocumentSearchResult {
  document_id: string;
  filename: string;
  relevance_score: number;
  snippet: string;
  chunk_index?: number;
  metadata?: Record<string, any>;
}

export interface DocumentSearchResponse {
  corpus_id: string;
  query: string;
  results: DocumentSearchResult[];
  total_results: number;
  processing_time_ms?: number;
}

export interface CorpusStats {
  corpus_id: string;
  document_count: number;
  total_size_bytes: number;
  total_chunks?: number;
  average_chunk_size?: number;
  embedding_dimensions?: number;
  last_updated: string;
  supported_file_types: string[];
}

// Component prop types
export interface RAGManagementProps {
  chatbotId: string;
  chatbotName: string;
}

export interface CorpusCardProps {
  corpus: Corpus;
  onEdit: (corpus: Corpus) => void;
  onDelete: (corpusId: string) => void;
  onSelect: (corpus: Corpus) => void;
  isSelected?: boolean;
}

export interface DocumentUploadProps {
  corpusId: string;
  onUploadComplete: (documents: Document[]) => void;
  onUploadProgress: (progress: UploadJob) => void;
  maxFileSize?: number;
  allowedFileTypes?: string[];
}

export interface DocumentLibraryProps {
  documents: Document[];
  isLoading?: boolean;
  searchQuery: string;
  onSearchChange: (query: string) => void;
  viewMode: 'grid' | 'list';
  onViewModeChange: (mode: 'grid' | 'list') => void;
  filterStatus: string;
  onFilterStatusChange: (status: string) => void;
  sortBy: string;
  onSortByChange: (sortBy: string) => void;
  onRefresh: () => void;
}

export interface CorpusCreatorProps {
  chatbotId: string;
  isOpen: boolean;
  onClose: () => void;
  onCorpusCreated: (corpus: Corpus) => void;
}

// Hook return types
export interface UseRAGCorporaResult {
  corpora: Corpus[];
  isLoading: boolean;
  error: string | null;
  createCorpus: (data: CreateCorpusRequest) => Promise<Corpus>;
  updateCorpus: (id: string, data: UpdateCorpusRequest) => Promise<Corpus>;
  deleteCorpus: (id: string) => Promise<void>;
  refreshCorpora: () => Promise<void>;
}

export interface UseDocumentUploadResult {
  uploadDocument: (file: File, corpusId: string, options?: Partial<DocumentUploadRequest>) => Promise<Document>;
  batchUpload: (files: File[], corpusId: string, options?: Partial<BatchUploadRequest>) => Promise<UploadJob>;
  validateFiles: (files: File[], corpusId: string) => Promise<UploadValidationResponse>;
  uploadProgress: Record<string, number>;
  isUploading: boolean;
  uploadErrors: string[];
}

export interface UseCorpusDocumentsResult {
  documents: Document[];
  isLoading: boolean;
  error: string | null;
  deleteDocument: (documentId: string) => Promise<void>;
  searchDocuments: (query: string) => Promise<DocumentSearchResult[]>;
  refreshDocuments: () => Promise<void>;
}

// Error types
export interface RAGError {
  message: string;
  code?: string;
  context?: Record<string, any>;
}

export interface APIErrorResponse {
  message: string;
  error_code?: string;
  timestamp: string;
  context?: Record<string, any>;
}

// Configuration types
export interface RAGConfig {
  enabled: boolean;
  corpus_id?: string;
  max_results?: number;
  min_relevance_score?: number;
  show_citations?: boolean;
}

// File type constraints
export interface FileTypeConfig {
  maxSize: number; // in bytes
  allowedTypes: string[]; // MIME types
  maxBatchSize: number; // total bytes for batch upload
  maxFilesPerBatch: number; // number of files
}

export const DEFAULT_FILE_CONFIG: FileTypeConfig = {
  maxSize: 50 * 1024 * 1024, // 50MB
  allowedTypes: [
    'application/pdf',
    'text/plain',
    'text/markdown', 
    'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
    'application/msword',
    'text/csv',
    'application/json'
  ],
  maxBatchSize: 500 * 1024 * 1024, // 500MB
  maxFilesPerBatch: 100
};

// UI state types
export interface RAGUIState {
  selectedCorpus?: Corpus;
  showCorpusCreator: boolean;
  showDocumentUpload: boolean;
  uploadProgress: Record<string, UploadJob>;
  searchQuery: string;
  searchResults: DocumentSearchResult[];
  viewMode: 'grid' | 'list';
  sortBy: 'created_at' | 'name' | 'size' | 'status';
  sortOrder: 'asc' | 'desc';
}